llama_index/node_parser/metadata_extraction.py

```diff
@@ -0,0 +1,250 @@
+            chunk_overlap=chunk_overlap,
+            callback_manager=callback_manager,
+        )
+        return cls(
+            text_splitter=token_text_splitter,
+            include_metadata=include_metadata,
+            include_prev_next_rel=include_prev_next_rel,
+            callback_manager=callback_manager,
+        )
+
+    # TODO: aget version of this
+    def get_nodes_from_documents(
+        self,
+        documents: Sequence[Document],
+    ) -> List[BaseNode]:
+        """Parse document into nodes.
+
+        Args:
+            documents (Sequence[Document]): documents to parse
+
+        """
+        event_id = self.callback_manager.on_event_start(
+            CBEventType.NODE_PARSING, payload={EventPayload.DOCUMENTS: documents}
+        )
+
+        all_nodes: List[BaseNode] = []
+        for document in documents:
+            nodes = get_nodes_from_document(
+                document,
+                self._text_splitter,
+                True,
+                include_prev_next_rel=self._include_prev_next_rel,
+            )
+            all_nodes.extend(nodes)
+        if any(["subsection" in mode for mode in self._extraction_modes]):
+            raise NotImplementedError(
+                "extracting subsection metadata is not yet implemented",
+                self._extraction_modes,
+            )
+        if any(["summary" in mode for mode in self._extraction_modes]):
+            # TODO: parallelize this with a `ThreadPoolExecutor` even for sync APIs
+            node_summaries = [
+                self._llm_predictor.predict(
+                    Prompt(
+                        template="""
+                        Here is the content of the section: {context_str}. 
+                        Summarize the key topics and entities of the section. 
+                        """
+                    ),
+                    context_str=node.text,
+                )[0]
+                for node in all_nodes
+            ]
+
+            # Extract node-level summary metadata
+            for i, node in enumerate(all_nodes):
+                if i > 0 and (
+                    MetadataExtractionMode.PREV_SUMMARY in self._extraction_modes
+                    or MetadataExtractionMode.PREV_NEXT_SUMMARY
+                    in self._extraction_modes
+                ):
+                    node.metadata["prev_section_summary"] = node_summaries[i - 1]
+                if i < len(all_nodes) - 1 and (
+                    MetadataExtractionMode.NEXT_SUMMARY in self._extraction_modes
+                    or MetadataExtractionMode.PREV_NEXT_SUMMARY
+                    in self._extraction_modes
+                ):
+                    node.metadata["next_section_summary"] = node_summaries[i + 1]
+                if MetadataExtractionMode.SELF_SUMMARY in self._extraction_modes:
+                    node.metadata["section_summary"] = node_summaries[i]
+
+        if MetadataExtractionMode.DOCUMENT_TITLE in self._extraction_modes:
```

Comment:
This heuristic is quite brittle and only works for certain types of documents.

We need a more representative set to refine for various cases.